vortex-expr/src/context.rs

```diff
@@ -0,0 +1,148 @@
+use std::any::Any;
+use std::iter;
+use std::sync::Arc;
+
+use itertools::Itertools;
+use vortex_array::aliases::hash_map::HashMap;
+use vortex_array::{Array, ArrayRef};
+use vortex_dtype::DType;
+use vortex_error::{VortexError, VortexExpect, VortexResult, vortex_err};
+
+use crate::IDENTITY_IDENTIFIER;
+
+pub type Identifier = Arc<str>;
+pub type ExprScope<T> = HashMap<Identifier, T>;
+
+#[derive(Clone, Default)]
+pub struct ValuesScope(ExprScope<ArrayRef>);
+#[derive(Clone, Default)]
+pub struct ValuesDTypeScope(ExprScope<DType>);
+#[allow(dead_code)]
+#[derive(Clone, Default)]
+pub struct VarsScope(ExprScope<Arc<dyn Any>>);
+
+#[derive(Clone, Default)]
+pub struct EvaluationContext {
+    array_len: usize,
```

Comment:
Probably hold an explicit `root: ArrayRef` just for easy access / no hash map when we don't have any vars